scripts/airgapped/get-all-images.py

```diff
@@ -0,0 +1,177 @@
+    script will also fail.
+    """
+    images = []
+    repo_dir = clone_repo(app[GH_REPO_KEY], app[GH_BRANCH_KEY])
+
+    logging.info(f"Executing repo's {GET_IMAGES_SH} script")
+    process = subprocess.Popen(["bash", "tools/get-images.sh"], cwd=repo_dir,
+                               stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    stdout, stderr = process.communicate()
+    if stderr:
+        raise ValueError("Script '%s' for charm '%s' had error logs: \n%s" %
+                         (GET_IMAGES_SH, app["charm"], stderr.decode("utf-8"))
+                         )
+    images = stdout.decode("utf-8").split("\n")
+
+    # cleanup
+    shutil.rmtree(repo_dir)
+    return images
+
+
+def get_dependency_app_images(app: dict) -> list[str]:
+    """
+    This function gets the images used by a dependency charm by:
+    1. Cloning the repo of the charm
+    2. Looking at its metadata.yaml for "upstream-source" keys
+    3. Delete the repo
+    """
+    images = []
+    repo_dir = clone_repo(app[GH_DEPENDENCY_REPO_KEY],
+                          app[GH_DEPENDENCY_BRANCH_KEY])
+    with open("%s/metadata.yaml" % repo_dir, 'r') as metadata_file:
+        metadata_dict = yaml.safe_load(metadata_file)
+
+    for _, rsrc in metadata_dict["resources"].items():
+        if "type" in rsrc and rsrc["type"] == "oci-image":
+            img = rsrc["upstream-source"]
+            logging.info("Found image %s" % img)
+            images.append(img)
+
+    # cleanup
+    shutil.rmtree(repo_dir)
+    return images
+
+
+def get_bundle_images(bundle_dict: dict) -> list[str]:
+    """Return a list of images used by a bundle"""
+    images = []
+
+    for app_name, app in bundle_dict["applications"].items():
+        logging.info(f"Handling app {app_name}")
+
+        # exclude repos we know don't have images
+        if app_name in EXCLUDE_REPOS:
+            logging.info("Ignoring charm %s", app["charm"])
+            continue
+
+        # Follow default image-gather process for dependency apps
+        if is_dependency_app(app):
+            logging.info("Dependency app '%s' with charm '%s'", app_name,
+                         app["charm"])
+            images.extend(get_dependency_app_images(app))
+            continue
+
+        # Ensure analytics app has necessary github repo/branch metadata
+        if not bundle_app_contains_gh_metadata(app):
+            raise KeyError(
+                "Application '%s' doesn't include expected gh metadata keys" %
+                app_name
+            )
+
+        images.extend(get_analytics_app_images(app))
+
+    return images
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        description="Gather all images from a bundle"
+    )
+    parser.add_argument("bundle")
+    parser.add_argument("--airgap-testing", action="store_true")
+
+    args = parser.parse_args()
+
+    bundle_dict = {}
+    with open(args.bundle, 'r') as file:
+        bundle_dict = yaml.safe_load(file)
```

Comment:
ACK and good idea to move all of the yaml loading and list parsing logic inside of the `get_bundle_images` function, as this would be logic we would need for any bundle we would like to process with this function.

Commit ad68c16e3b1ae120cdbbf40c45b30507c02d81f2

On the proposed changes though, I don't think we can pass the path string to the `safe_load` directly as it needs to get a `stream` object as input. Do you think I'm missing something?
https://pyyaml.org/wiki/PyYAMLDocumentation

Reply:
> On the proposed changes though, I don't think we can pass the path string to the safe_load directly as it needs to get a stream object as input. Do you think I'm missing something?

Sorry, I was messing this with another method, but you can anyway do `bundle_object = yaml.safe_load(Path(args.bundle).read_text())` and you avoid the context manager. The context manager is useful for when you want to read the contents of the file and perhaps do something about them, perhaps edit or parse in a certain way, and `Path.read_text()` is best for cases where you just want to pass the file object to something else, in this case `yaml.safe_load()`.

You have to make sure the `args.bundle` exists, still.

Reply:
Hmm I hadn't seen this pattern else where, is it a common one we follow in our charms or other places in the org?

Reply:
Also, I'm not in sync with the logic on not doing it via the context manager. Why would we want to avoid the context manager?

The other way around seems like a more complicated path to opening a file. I can't find other projects following the `Path.read_text()` approach

Reply:
> The other way around seems like a more complicated path to opening a file. I can't find other projects following the Path.read_text() approach

What do you mean it is more complicated? For reference, we use this in chisme.

> Why would we want to avoid the context manager?

Please check https://github.com/canonical/bundle-kubeflow/pull/1038#discussion_r1741656221

Reply:
Took a slightly more detailed look.

Looks like the `read_text()` method reads the contents and closes the file, so it's indeed equivalent and can be replaced in case we only want the contents of the file (to avoid using a context manager).

Learned something new, thanks!

d20c82b